dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,670 @@
 
-	// Requests allows the user to specify the minimum resource
-	// requirements that must be satisfied across all devices allocated for
-	// this claim.
+	// Future extension, not part of the current design:
+	// A CEL expression which compares different devices and returns
+	// true if they match.
 	//
-	// All drivers can the support "devices", resource, which represents
-	// the minimum number of devices to allocate.
+	// Because it would be part of a one-of, old schedulers will not
+	// accidentally ignore this additional, for them unknown match
+	// criteria.
 	//
-	// Other resource types are driver-specific. The values in `requests`
-	// represent the aggregate across all devices, not the per-device
-	// values.
+	// matcher string
+}
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the result of that was.
+type ResourceClaimStatus struct {
+	// ControllerName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started. It is empty when the claim was
+	// allocated through structured parameters,
 	//
-	// If not set, requests.devices will default to 1.
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
 	//
 	// +optional
-	Requests map[string]resource.Quantity `json:"requests,omitempty"`
+	ControllerName string `json:"controllerName"`
 
-	// MatchAttributes allows specifying a constraint within a set of
-	// chosen devices, without having to explicitly specify the value of
-	// the constraint.  For example, this allows constraints like "all
-	// devices must be the same model", without having to specify the exact
-	// model. We may be able to use this for some basic topology
-	// constraints too, by representing the topology as device attributes.
+	// Allocation is set once the claim has been allocated successfully.
+	// +optional
+	Allocation *AllocationResult `json:"allocation,omitempty" protobuf:"bytes,2,opt,name=allocation"`
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. A Pod which references a ResourceClaim which is not
+	// reserved for that Pod will not be started.
 	//
-	// Currently, these are just strings. However, we could make them
-	// structs, and include required vs preferred matches. Required matches
-	// would fail if not met, where as preferred would lower the score if
-	// not met. We could even allow low/medium/high priority and adjust the
-	// score differently for each.
+	// There can be at most 32 such reservations. This may get increased in
+	// the future, but not reduced.
 	//
+	// +listType=map
+	// +listMapKey=uid
+	// +patchStrategy=merge
+	// +patchMergeKey=uid
 	// +optional
-	// +listType=atomic
-	MatchAttributes []string `json:"matchAttributes,omitempty"`
+	ReservedFor []ResourceClaimConsumerReference `json:"reservedFor,omitempty" protobuf:"bytes,3,opt,name=reservedFor" patchStrategy:"merge" patchMergeKey:"uid"`
```

Comment:
What is the use case for this? I think right now, we need to be downscoping things to a minimal set of most critical use cases.

Reply:
Sharing of a claim between pods.

Reply:
I still don't know why we need this. The pod spec points at the claim. Why do we need a back reference here? Once a pod gets created that is using the claim, it could add an ownerRef to the claim.

Reply:
The back reference is there to prevent deallocating the claim while it is in use. An ownerRef doesn't achieve that.

Reply:
I'm still not following. I don't think it should block the PR but I would like to understand this better.

Reply:
Patrick, you are probably the only person on the planet who has this KEP entierly in their brain.  :)

"""
In a cluster with multiple scheduler instances, two pods might get scheduled concurrently by different schedulers. When they reference the same ResourceClaim which may only get used by one pod at a time, only one pod can be scheduled.

Both schedulers try to add their pod to the claim.status.reservedFor field, but only the update that reaches the API server first gets stored. The other one fails with a conflict error and the scheduler which issued it knows that it must put the pod back into the queue, waiting for the ResourceClaim to become usable again.
"""

IOW: a claim might be allocated and not-in-use or allocated and in-use, and the difference matters

Reply:
I don't think this should be in scope. This is effectively "platform-mediated sharing".

Reply:
@thockin: I take that as a suggestion to add that to the API documentation. This goes back to our discussion about what should or should not be in those docs. I'm fine with adding it, so I pushed directly to the branch.

@johnbelamaric: limiting to one consumer is only one aspect of it. The other, more common usage is to avoid race conditions when multiple pods are allowed to use it: deallocation must not happen when in use. Added also to the doc comment.